### DISCUSSION TITLE
Custom data with PayPal Standard

### DISCUSSION THREAD

**Discussion by Genevieve-K** [impress-org/givewp#6690] [Developer Q&A]
Hi. I have a problem when the donor uses PayPal Standard. For my cart, I need some data after payment validation, I store these data in $_REQUEST. For PayPal Standard, when the confirmation page is displayed after the payment, the initial $_REQUEST is not saved. 
I see that a $_REQUEST is constructed in the `generateSecureGatewayRouteUrl(string $gatewayMethod, int $donationId, array $args = [])` function, in file PaymentGateway.php. I think I could add data in this $_REQUEST with the filter
`handleCreatePayment(Donation $donation)` :  `$gatewayData = apply_filters(
                "givewp_new_payment_{$donation->gatewayId}_gateway_data",
                null,
                $donation
            );`
$gatewayData is passed as an argument to the `createPayment()` function. But in file PayPalStandard.php, in `createPayment()`, this variable  $gatewayData is not added to $args. 
Could you add it?

**Comment by jonwaldstein** [comment]
Hey @Genevieve-K 

You are correct, gateways have the ability to use the `$gatewayData` param and filter to send data to itself from the form.  In the latest version of GiveWP, PayPalStandard does indeed have that param in `createPayment().` 

Can you help me understand what exactly you need to do after payment validation?  What data do you need and what do you need to do with that data?

**Reply by Genevieve-K** [reply]
I developped a cart. The donor can make several donations ans pay them together. But if he makes a one-off donation and a recurring donation, he must pay them separately. I remove the donations from the cart when the payment is done, wheter it is pending or complete. I added an input in $_REQUEST to know the donations to remove from the cart.

**Reply by jonwaldstein** [reply]
Okay, I think what you are trying to do would require overriding a few parts of the gateway class in order to send specific request params along.  It might actually be easier for you to use sessions to retrieve the donation / subscription ID's and check their statuses that way.
